`ConvertTo-Csv`: Quote fields with quotes and newlines when using `-UseQuotes AsNeeded`

lselden opened the pull request:
Fix for issue #9284 - Escape fields that contain quotes and newlines, not just those that contain the delimiter

# PR Summary

This PR fixes the case of CSV fields that include double quotes or newlines in their values. If a Column value contains `"`,`\r`,`\n` or the delimiter then it will be marked as needing to be surrounded by quotes.

This follows [RFC-4180 rules](https://datatracker.ietf.org/doc/html/rfc4180#section-2) as to what characters trigger a field to be quoted (except it allows for different delimiters than the comma). This also follows how `ConvertFrom-Csv` parses input CSV files.

## PR Context

`-UseQuotes AsNeeded` currently only looks for delimiters, which means that the field `Hello,World` will be surrounded in quotes, however the field `Hello"World` would not be surrounded in quotes, causing this kind of embarassing slip-up:

```powershell
$testObj = [pscustomobject]@{ Greet = "Hello"; Me = "`""; Here = "World" };
# Greet Me Here
# ----- -- ----
# Hello "  World

$str = $testObj | convertto-csv -UseQuotes AsNeeded
# Greet,Me,Here
# Hello,",World  <-- Should be 'Hello,"""",World'

$str | convertfrom-csv
# Greet Me     Here
# ----- --     ----
# Hello ,World
```

Newlines are another risk, since they're interpreted as completely different rows.

```powershell
$testObj = [pscustomobject]@{ FirstColumn = "Hello"; SecondColumn = "World`r`nGoodbye" }
FirstColumn,SecondColumn
Hello,World
Goodbye
```

**Performance Caveat:** This version is searching for the existence of 4 characters (using `string.IndexOfAny(chars)` instead of the original 1 (`string.Contains(char)`. Based on my benchmarks this change does have a *miniscule* increase in time for that single operation. However, in my naïve in-memory benchmarks against a 100MB CSV file, using `-UseQuotes AsNeeded` was slightly faster than `-UseQuotes Always` or `-QuoteFields`. My guess is that I didn't need to allocate as much memory for those unneeded quotes.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [X] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [X] Issue filed: [7796](https://github.com/MicrosoftDocs/PowerShell-Docs/issues/7796)
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [X] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed: <!-- Number/link of that issue here -->

lukeselden committed:
Escape quotes and newlines with -UseQuotes AsNeeded

Fix for issue #9284 - Escape fields that contain quotes and newlines, not just those that contain the delimiter

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/CsvCommands.cs:
```diff
@@ -967,7 +970,7 @@ internal string ConvertPropertyNamesCSV(IList<string> propertyNames)
                             break;
                         case BaseCsvWritingCommand.QuoteKind.AsNeeded:
-                            if (propertyName.Contains(_delimiter))
+                            if (propertyName.IndexOfAny(_escapeCharacters) != -1)
```
Can we use `public static unsafe int IndexOfAny(ref char searchStart, char value0, char value1, char value2, char value3, int length)` to avoid allocation of `_escapeCharacters`?

lselden replied:
@iSazonov That sounds awesome! However, I'll admit I'm a Typescript developer, not a C# developer. Is that a [dotnet private method call](https://source.dot.net/#System.Private.CoreLib/SpanHelpers.Char.cs,1235)? I'm open to suggestions

iSazonov replied:
I see public https://source.dot.net/#System.Private.CoreLib/SpanHelpers.Char.cs,985

lselden replied:
I couldn't find any way to reference that method - I couldn't find any public references to it in github/code, and even tried converting the string to a [Span](https://docs.microsoft.com/en-us/dotnet/api/system.memoryextensions.indexofany?view=net-6.0#System_MemoryExtensions_IndexOfAny__1_System_Span___0____0___0___0_) but couldn't get it to work. It certainly seems preferable if you can show me how to use it / commit yourself.

FWIW I do see a [number of references](https://github.com/search?p=2&q=org%3APowerShell+indexofany&type=Code) to `string.IndexOfAny()` in this repo, but they aren't in this kind of inner loop situation where performance is a concern.

iSazonov replied:
Public is here https://source.dot.net/#System.Private.CoreLib/MemoryExtensions.cs,692

lselden replied:
Okay, [updated accordingly](https://github.com/PowerShell/PowerShell/pull/15765/commits/4832f43b91128e700b131a969c052e800ef6549f). I had to take out `\r` detection but I don't think that will matter since that character should always be followed by `\n`.

iSazonov replied:
> I had to take out `\r` detection

Why?

iSazonov reviewed (commented):
Please add a test with new line chars.

lukeselden committed:
add additional -UseQuotes AsNeeded tests

Added additional tests for different character scenarios that should be handled

lselden commented:
Re: Testing - Added [additional tests](https://github.com/PowerShell/PowerShell/pull/15765/commits/c712df9199858563923f532c6305bf3232759053) for detecting delimiters, newlines and quotes. I did put the test objects inside the `It` block, since they were only used per-test - let me know if they should be hoisted up to the Context's `BeforeAll` section.

lukeselden committed:
Code Quality cleanup per codefactor review

lukeselden committed:
inline AsNeeded character detection

Cast string as ReadOnlySpan to avoid allocation of char array

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertTo-Csv.Tests.ps1:
```diff
@@ -159,5 +159,42 @@ Describe "ConvertTo-Csv" -Tags "CI" {
+
+            $result[0] | Should -BeExactly "`"FirstColumn`"rSecond,Column"
+            $result[1] | Should -BeExactly "Hello,r`"World`""
+        }
+        It "UseQuotes AsNeeded Escapes Newlines" {
```
```suggestion
        }

        It "UseQuotes AsNeeded Escapes Newlines" {
```

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertTo-Csv.Tests.ps1:
```diff
@@ -159,5 +159,42 @@ Describe "ConvertTo-Csv" -Tags "CI" {
+
+            $result[0] | Should -BeExactly "`"First`nColumn`""
+            $result[1] | Should -BeExactly "`"Hello`nWorld`""
+        }
+        It "UseQuotes AsNeeded Escapes Quotes" {
```
```suggestion
        }

        It "UseQuotes AsNeeded Escapes Quotes" {
```

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertTo-Csv.Tests.ps1:
```diff
@@ -159,5 +159,42 @@ Describe "ConvertTo-Csv" -Tags "CI" {
+
+            $result[0] | Should -BeExactly "`"First`"`"Column`""
+            $result[1] | Should -BeExactly "`"`"`"Hello`"`" World`""
+        }
+
```
Please remove the empty extra line.

lselden committed:
Apply suggestions from code review

Co-authored-by: Ilya <darpa@yandex.ru>
